sus/mem/macros.h

```diff
@@ -0,0 +1,229 @@
+  friend struct ::sus::mem::__private::RelocatableTag;                       \
+  static constexpr bool SusUnsafeTrivialRelocate = is_trivially_reloc
+
+/// Mark a class as unconditionally trivially relocatable, without any
+/// additional assertion to help verify correctness.
+///
+/// Generally, prefer to use sus_class_trivially_relocatable() with
+/// all field types passed to the macro.
+/// To additionally allow the class to be passed in registers, the class can be
+/// marked with the `[[clang::trivial_abi]]` attribute.
+///
+/// Use the [`TriviallyRelocatable`]($sus::mem::TriviallyRelocatable) concept to
+/// determine if a type is trivially relocatable, and to test with static_assert
+/// that types are matching what you are expecting. This allows containers to
+/// optimize their implementations when relocating the type in memory.
+///
+/// | Macro | Style |
+/// | ----- | ----- |
+/// | [`sus_class_trivially_relocatable`]($sus_class_trivially_relocatable) | **asserts** all param types are trivially relocatable |
+/// | [`sus_class_trivially_relocatable_if_types`]($sus_class_trivially_relocatable_if_types) | is **conditionally** trivially relocatable if all param types are |
+/// | [`sus_class_trivially_relocatable_if`]($sus_class_trivially_relocatable_if) | is **conditionally** trivially relocatable if the condition is true |
+/// | [`sus_class_trivially_relocatable_unchecked`]($sus_class_trivially_relocatable_unchecked) | is trivially relocatable without any condition or assertion |
+///
+/// # Example
+/// ```
+/// struct S {
+///   Thing<i32> thing;
+///   i32 more;
+///
+///   sus_class_trivially_relocatable_unchecked(unsafe_fn);
+/// };
+/// ```
+#define sus_class_trivially_relocatable_unchecked(unsafe_fn)          \
+  static_assert(std::is_same_v<decltype(unsafe_fn),                   \
+                               const ::sus::marker::UnsafeFnMarker>); \
+  template <class SusOuterClassTypeForTriviallyReloc>                 \
+  friend struct ::sus::mem::__private::RelocatableTag;                \
+  static constexpr bool SusUnsafeTrivialRelocate = true
+
+/// Mark a class field as never being a specific value, often a zero, after a
+/// constructor has run and before the destructor has completed. This allows
+/// querying if a class is constructed in a memory location, since the class is
+/// constructed iff the value of the field is not the never-value.
+///
+/// The named field can be compared to the `never_value` to determine if the
+/// object is constructed. The field must be set to the `destroy_value` just
+/// prior to destruction. The latter is meant to help the destructor be a no-op
+/// when the type is in a never-value state, if the never-value would be read in
+/// the destructor.
+///
+/// The macro includes `private:` which changes the class definition visibility
+/// to private.
+#define sus_class_never_value_field(unsafe_fn, T, field_name, never_value,     \
```

Comment:
can we put this one in never_value_macros.h, and the others in relocate_macros.h?

Reply:
Done.